src/main/java/com/feedzai/commons/sql/abstraction/engine/BlobEncoder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2015 Feedzai
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.feedzai.commons.sql.abstraction.engine;
+
+import com.feedzai.commons.sql.abstraction.util.InitiallyReusableByteArrayOutputStream;
+import com.google.common.io.ByteStreams;
+
+import java.io.*;
+
+/**
+ * Logic to encode java values into blob values and vice-versa. BLOB values are
+ * stored with a byte prefix, the encoder id, followed by the encoded value. The
+ * encoding to use is the most efficient for the type of the java value to store.
+ *
+ * @author  Paulo Leitao (paulo.leitao@feedzai.com)
+ * @since 2.1.5
+ */
+public class BlobEncoder {
+
+    /**
+     * Prefix for objects encoded with java serialization, set with the same value as the first byte
+     * of STREAM_MAGIC in the java serialization protocol (java serialized objects always start by AC ED).
+     * In order to maintain backward compatibility, this prefix is part of the serialized value.
+     */
+    private static final int ID_JAVA_ENCODER = 0xac;
+    /**
+     * Prefix for objects not serialized.
+     */
+    private static final int ID_NULL_ENCODER = 0x00;
+
+    /**
+     * Size of blob buffer.
+     */
+    private int blobBufferSize;
+    /**
+     * The reusable initial byte buffer for blob values, allocated only if needed.
+     */
+    private byte[] reusableByteBuffer;
+
+    /**
+     * Create new encoder.
+     *
+     * @param blobBufferSize  Encoder internal buffer size, the maximum size of a blob value.
+     */
+    public BlobEncoder(int blobBufferSize) {
+        this.blobBufferSize = blobBufferSize;
+    }
+
+    /**
+     * Converts an object to a byte array representation in a BLOB value.
+     *
+     * @param val The object to convert.
+     * @return The byte array representation of the object.
+     * @throws IOException If the buffer is not enough to make the conversion.
+     */
+    public final synchronized byte[] encode(Object val) throws IOException {
```

Comment:
we should replace synchronized with a ReentrantLock and use it only in the `else` part since that is where we need it due to the reusable byte buffer